minidump-processor/src/stackwalker/amd64.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright 2015 Ted Mielczarek. See the COPYRIGHT
+// file at the top-level directory of this distribution.
+
+// Note since x86 and Amd64 have basically the same ABI, this implementation
+// is written to largely erase the details of the two wherever possible,
+// so that it can be copied between the two with minimal changes. It's not
+// worth the effort to *actually* unify the implementations.
+
+use crate::process_state::{FrameTrust, StackFrame};
+use crate::stackwalker::unwind::Unwind;
+use minidump::format::CONTEXT_AMD64;
+use minidump::{
+    MinidumpContext, MinidumpContextValidity, MinidumpMemory, MinidumpModuleList,
+    MinidumpRawContext,
+};
+use std::collections::HashSet;
+
+type Pointer = u64;
+const POINTER_WIDTH: Pointer = 8;
+const INSTRUCTION_REGISTER: &str = "rip";
+const STACK_POINTER_REGISTER: &str = "rsp";
+const FRAME_POINTER_REGISTER: &str = "rbp";
+
+fn get_caller_by_frame_pointer(
+    ctx: &CONTEXT_AMD64,
+    valid: &MinidumpContextValidity,
+    _trust: FrameTrust,
+    stack_memory: &MinidumpMemory,
+    _modules: &MinidumpModuleList,
+) -> Option<StackFrame> {
+    match valid {
+        MinidumpContextValidity::All => {}
+        MinidumpContextValidity::Some(ref which) => {
+            if !which.contains(FRAME_POINTER_REGISTER) {
+                return None;
+            }
+        }
+    }
+
+    let last_bp = ctx.rbp;
+    // Assume that the standard %bp-using x64 calling convention is in
+    // use.
+    //
+    // The typical x86 calling convention, when frame pointers are present,
+    // is for the calling procedure to use CALL, which pushes the return
+    // address onto the stack and sets the instruction pointer (%ip) to
+    // the entry point of the called routine.  The called routine then
+    // PUSHes the calling routine's frame pointer (%bp) onto the stack
+    // before copying the stack pointer (%esp) to the frame pointer (%bp).
+    // Therefore, the calling procedure's frame pointer is always available
+    // by dereferencing the called procedure's frame pointer, and the return
+    // address is always available at the memory location immediately above
+    // the address pointed to by the called procedure's frame pointer.  The
+    // calling procedure's stack pointer (%esp) is 8 higher than the value
+    // of the called procedure's frame pointer at the time the calling
+    // procedure made the CALL: 4 bytes for the return address pushed by the
+    // CALL itself, and 4 bytes for the callee's PUSH of the caller's frame
+    // pointer.
+    //
+    // %ip_new = *(%bp_old + ptr)
+    // %sp_new = %bp_old + ptr
+    // %bp_new = *(%bp_old)
+
+    let caller_ip = stack_memory.get_memory_at_address(last_bp as u64 + POINTER_WIDTH as u64)?;
+    let caller_bp = stack_memory.get_memory_at_address(last_bp as u64)?;
+    let caller_sp = last_bp + POINTER_WIDTH * 2;
```

Comment:
The above comment assumes 4 bytes per pointer, but here 8 bytes are used (since this is a 64bit architecture). I assume this is because the comment refers to the situation on x86? It might be worthwhile to change the byte values in the comment or mention why they don't match the code below.